designs/2022-async-parse/README.md

```diff
@@ -0,0 +1,258 @@
+- Repo: eslint/eslint
+- Start Date: 2022-03-17
+- RFC PR: (leave this empty, to be filled in later)
+- Authors: Milos Djermanovic
+
+# Support async parsing
+
+## Summary
+
+ESLint now requires that `parse` methods of custom parsers return `AST` synchronously. This RFC proposes to allow the `parse` methods of custom parsers to return `Promise<AST>`. Similarly, the `parseForESLint` methods of custom parsers would be allowed to return `Promise<ParseResult>` instead of `ParseResult`.
+
+## Motivation
+
+Per the [original issue](https://github.com/eslint/eslint/issues/15475), this feature will allow for the use of ESLint with custom parsers written in languages that compile to Wasm or other binary formats.
```

Comment:
This motivation is speculative since there is no such asynchronous parser for now. SWC is a candidate but it has performance problem passing AST to JS. https://github.com/swc-project/swc/issues/2175

Reply:
I’d say anticipatory rather than speculative. People won’t write parsers that require async loading until ESLint supports async loading. We want to allow for the possibility sooner rather than later.

Reply:
the import plugin would likely benefit massively from an async parser, that can make parsing multiple files concurrent across the whole codebase.